Jobtime / Playtime Tracking

TheDarkElites opened the pull request:
## About the Pull Request

This will allow us to track player's time in jobs and then store that information on our server's database.
There is also now a playtime tab in the player panel to check a player's playtime for admins.
Roles can now be time-restricted.

P.S.: The reason the last pull request was closed was because it was a mess and I needed to start from scratch.

**Do Not Merge**

## Why It's Good For The Game

less LRP more HRP

## Current Time Requirements 
_(This is a baseline so tusurpta has something to work with when adjusting time requirements, and may not be the final requirements in game.)_

![image](https://github.com/Foundation-19/Foundation-19/assets/73414180/2b615b5b-cb9f-486f-b4d5-2186e665f95b)
![image](https://github.com/Foundation-19/Foundation-19/assets/73414180/3cd6a7da-b2e6-49e0-8df7-59ada54e075e)


~~You can also find it in text form as follows:~~ Now Outdated, refer to picture above.

```
Site Director: 20 hrs Command
HOP: 2 hrs Command, 8hrs service, 8hrs bureacrat
Comms Officer: 10hrs Engineering, 10hrs Security, Comms tech 15hrs
Comms Tech: 5hrs Engineering
Internal Tribunal Department Officer: 10 hrs command, 15 hrs security, 2hrs bureacrat
Ethics Liaison: Command 3hrs, Bureacrat 8hrs
----------------------------------
Junior Engineer: Class D 30min
Engineer: 5hrs Junior Engineer
Senior Engineer: 10hrs Engineer, 12hrs Engineering
Containment Engineer: 10hrs Senior Engineer, 20hrs engineering
Chief Engineer: 10hrs Containment Engineer, 25hrs engineering
----------------------------------
EMT: Doctor 5hrs, Medical 10hrs
Surgeon: Doctor 8hrs, Medical 15hrs
Doctor: DClass 30min
Psychiatrist: Doctor 1hr
Chemist: DClass 30min
CMO: 2hrs EMT, 5hrs Surgeon, 10hrs Medical Doctor, 25hrs medical
----------------------------------
Logistics Officer: 5hr Logistics Specialist
Logistics Specialist: Class D 30min
----------------------------------
Archavist: 5hr Office Worker
----------------------------------
Junior Researcher: Class D 30min
Junior roboticist: Class D 30min
Researcher: 5hrs Junior Researcher
Roboticist: 5hrs Junior Roboticist
Senior Researcher: 10hrs Researcher, 12hrs Science
Senior Roboticist: 10hrs Roboticist, 10hrs Science
RD: 10hrs Senior Researcher, 25hrs Science
----------------------------------
Guard Commander: 5hr as each zone commander, 30hrs as security
Zone Commanders: 10hrs LCZ/HCZ/EZ(for their respective zone) Seargent/Senior Agent, 25hrs as LCZ/HCZ/EZ(for their respective zone)
Seargent/Senior Agent: 10hrs as LCZ/HCZ/EZ(for their respective zone) Guard/Agent, 12hrs as LCZ/HCZ/EZ(for their respective zone)
LCZ Guard: Class D 1hr
EZ Guard: LCZ 5hr
HCZ Guard: EZ 5hr
----------------------------------
AIC: 20hrs robot
```

## Changelog

:cl:
add: Jobtime Tracking and Timelocks for certain roles
server: new SQL table for role_time
code: goirep no longer has the COM flag, and is replaced with the new REP flag
admin: player panel now contains playtime tab
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

<!--
Examples were changelog entries are optional/not typically required but encouraged:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

As a courtesy, for ported PRs, please include if you have the blessing of the author. While not required, we encourage basic decency in porting others' work. It is also sufficient to note that the author has not expressed displeasure at the idea of their work getting ported.
Please refrain from porting works of authors that have expressed displeasure in having their work ported, thank you.
-->

TheDarkElites committed:
Initial framework implementation for job  time tracking

jobtime subsystem + SQL integration + and rudementary jobtime tracking

TheDarkElites commented:
@Tsurupeta add do not merge label please

Tsurupeta commented:
Nobody's merging drafts anyways

TheDarkElites committed:
finalizes tracking implementation and subsystem

TheDarkElites committed:
SCP time tracking and jobtime datum framework + playtime query framework

TheDarkElites committed:
removes unnecessary var/

TheDarkElites committed:
adds calculation for sub-categories and departments

TheDarkElites committed:
fix checks

TheDarkElites committed:
Merge branch 'dev' into timelocks-(for-real-this-time)

TheDarkElites committed:
Adds BUR, LCZ, HCZ, and EZ flags to appropriate jobs

TheDarkElites committed:
spelling mistake fix

TheDarkElites committed:
Crew -> Staff

TheDarkElites committed:
Adds proper antag tracking and finishes adding flags

TheDarkElites committed:
fixes typo

TheDarkElites committed:
Adds proper initialization for jobtime datum

TheDarkElites committed:
requirement checking and requirements framework + admin tracking and admin override

TheDarkElites committed:
Timelock implementation

TheDarkElites committed:
Adds appropriate UI in occupations for timelocks and adds backend support for guest users

TheDarkElites committed:
adds in timelock config + admin verb to toggle timelocks. Also adds in jobtime update cooldown to avoid unnecessary SQL queries

TheDarkElites committed:
Adds playtime tab to player panel + fixes indentation in admin_verbs.dm

TheDarkElites committed:
Adds req checks for latejoin menu

TheDarkElites committed:
adds in null saftey to get jobtime procs

TheDarkElites committed:
Finalizes See Playtime verbs for players to check their playtime

TheDarkElites committed:
maybe finally fixes checks?

TheDarkElites committed:
Adds requirements for jobs and adds proper branch support for REP flag

TheDarkElites committed:
Cleanup

TheDarkElites committed:
Switches jobtime to fire every five minutes and to record time in minutes

TheDarkElites committed:
Grammar fix

TheDarkElites committed:
Adds SQL file for adding role_time table

TheDarkElites commented:
Okay this should be good for review.

TheDarkElites commented:
@Tsurupeta When you review this, also look over the current set of time requirements and make any adjustments you see fit.

TheDarkElites committed:
adds playtime config to example config file

TheDarkElites committed:
changes jobtime wait time to use macros

TheDarkElites committed:
Autodoc fixes

TheDarkElites committed:
use_jobtime_tracking -> use_timelocks

TheDarkElites committed:
fixes up toggle time locks verb

Tsurupeta commented on code/controllers/subsystems/jobs.dm:
```diff
@@ -10,6 +10,11 @@ var/const/SUP               =(1<<8)
 var/const/SPT               =(1<<9)
 var/const/EXP               =(1<<10)
 var/const/ROB               =(1<<11)
+var/const/LCZ               =(1<<12)
+var/const/ECZ               =(1<<13)
+var/const/HCZ               =(1<<14)
+var/const/BUR				=(1<<15)
+var/const/REP				=(1<<16)
```
replace all of this with defines.........

TheDarkElites replied:
That would mean throwing out the already existing department flags system (and everything that relies on it), unless we can do flags with defines somehow?

Tsurupeta commented on code/controllers/subsystems/jobtime.dm:
```diff
@@ -0,0 +1,97 @@
+	var/last_fired
+
+
+/datum/controller/subsystem/jobtime/Initialize()
+	if(!SSdbcore.IsConnected())
+		log_world("Jobtime tracking cannot function when DB is disconnected!")
+	last_fired = world.time
```
And? Do you set canfire or what? You get the log and it continues working.

Tsurupeta commented on code/controllers/subsystems/jobtime.dm:
```diff
@@ -0,0 +1,97 @@
+
+//Client Procs
+
+/client/proc/update_jobtime_list(minutes)
```
Does it all need to be defined on client?

Tsurupeta commented on code/controllers/subsystems/jobtime.dm:
```diff
@@ -0,0 +1,97 @@
+			"minutes" = jvalue)))
+	addtimer(CALLBACK(SSjobtime, /datum/controller/subsystem/jobtime/proc/update_jobtime_db),20,TIMER_OVERRIDE|TIMER_UNIQUE)
+
+/client/proc/get_jobtime_list_db()
```
Along with this.

TheDarkElites committed:
Clean up

fixes stuff as directed by review

TheDarkElites committed:
fixed minor oopsie

TheDarkElites commented:
![image](https://github.com/Foundation-19/Foundation-19/assets/73414180/a43f024f-a3d6-47da-aee7-8cf60a79a5d3)
Since Tusurpta didn't actually sign off on it here, ill just put this as his review.

Tsurupeta reviewed (dismissed):
Looks fine

TheDarkElites committed:
Merge branch 'dev' into timelocks-(for-real-this-time)

77seven77seven commented:
wait wont this trap everyone into playing only class-D?

TheDarkElites commented:
> wait wont this trap everyone into playing only class-D?

Not everyone, just new players, and only for 30 minutes. However, I'm going to go over the reqs list again and modify it to be cleaner. I'm probably only gonna keep the D-Class requirement for security.

77seven77seven commented:
> > wait wont this trap everyone into playing only class-D?
> 
> Not everyone, just new players, and only for 30 minutes. However, I'm going to go over the reqs list again and modify it to be cleaner. I'm probably only gonna keep the D-Class requirement for security.

ngl the D-class requirement for security is a good idea, this is going to make sec treat class-D better because they have to deal with sec doing the same shit to them

TheDarkElites commented:
> > > wait wont this trap everyone into playing only class-D?
> > 
> > 
> > Not everyone, just new players, and only for 30 minutes. However, I'm going to go over the reqs list again and modify it to be cleaner. I'm probably only gonna keep the D-Class requirement for security.
> 
> ngl the D-class requirement for security is a good idea, this is going to make sec treat class-D better because they have to deal with sec doing the same shit to them

Yes that is the point.

77seven77seven commented:
> > > > wait wont this trap everyone into playing only class-D?
> > > 
> > > 
> > > Not everyone, just new players, and only for 30 minutes. However, I'm going to go over the reqs list again and modify it to be cleaner. I'm probably only gonna keep the D-Class requirement for security.
> > 
> > 
> > ngl the D-class requirement for security is a good idea, this is going to make sec treat class-D better because they have to deal with sec doing the same shit to them
> 
> Yes that is the point.

great minds think alike :trollface:

TheDarkElites committed:
updates reqs to be cleaner and removes some unnecessary D-Class reqs

TheDarkElites committed:
Merge branch 'timelocks-(for-real-this-time)' of https://github.com/TheDarkElites/Foundation-19 into timelocks-(for-real-this-time)

TheDarkElites committed:
fixes init order to avoid null calls

TheDarkElites committed:
fixes runtimes

TheDarkElites committed:
fixes checks

TheDarkElites committed:
makes this a little better

TheDarkElites committed:
Merge branch 'dev' into timelocks-(for-real-this-time)

TheDarkElites committed:
fixes more runtimes

TheDarkElites committed:
updates config to actually work now

Timelocks by default disabled unless USE_TIMELOCKS is in config file as shown in example folder.

TheDarkElites committed:
Merge branch 'timelocks-(for-real-this-time)' of https://github.com/TheDarkElites/Foundation-19 into timelocks-(for-real-this-time)

TheDarkElites committed:
Merge branch 'dev' into timelocks-(for-real-this-time)

TheDarkElites committed:
finalizes requirements for the time being

TheDarkElites committed:
Merge branch 'dev' into timelocks-(for-real-this-time)
